SecurityPkg/Tcg/Tcg2Config/Tcg2ConfigFfaPeim.c

```diff
@@ -0,0 +1,213 @@
+  if ((MajorVer != TPM2_FFA_SERVICE_MAJOR_VERSION) || (MinorVer < TPM2_FFA_SERVICE_MINOR_VERSION)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Wrong Tpm Ffa Interface Version...: v%d.%d\n",
+      __func__,
+      MajorVer,
+      MinorVer
+      ));
+    return EFI_UNSUPPORTED;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  The entry point for Tcg2 configuration driver.
+
+  @param  FileHandle  Handle of the file being invoked.
+  @param  PeiServices Describes the list of possible PEI Services.
+**/
+EFI_STATUS
+EFIAPI
+Tcg2ConfigFfaPeimEntryPoint (
+  IN       EFI_PEI_FILE_HANDLE  FileHandle,
+  IN CONST EFI_PEI_SERVICES     **PeiServices
+  )
+{
+  EFI_STATUS              Status;
+  UINTN                   Size;
+  UINT16                  PartId;
+  EFI_FFA_PART_INFO_DESC  *TpmPartInfo;
+  VOID                    *TxBuffer;
+  UINT64                  TxBufferSize;
+  VOID                    *RxBuffer;
+  UINT64                  RxBufferSize;
+  UINT32                  Count;
+  BOOLEAN                 NeedReleaseRxTxBuffer;
+
+  DEBUG ((DEBUG_INFO, "%a\n", __func__));
+
+  NeedReleaseRxTxBuffer = FALSE;
+
+  Status = ArmFfaLibPartitionIdGet (&PartId);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to get partition-id. Status: %r\n", Status));
+
+    goto ErrorHandler;
+  }
+
+  Status = ArmFfaLibGetRxTxBuffers (
+             &TxBuffer,
+             &TxBufferSize,
+             &RxBuffer,
+             &RxBufferSize
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to get Rx/Tx Buffer. Status: %r\n", Status));
+    goto ErrorHandler;
+  }
+
+  Status = ArmFfaLibPartitionInfoGet (
+             &gTpm2ServiceFfaGuid,
+             FFA_PART_INFO_FLAG_TYPE_DESC,
+             &Count,
+             (UINT32 *)&Size
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to get Tpm2 partition info. Status: %r\n", Status));
+    goto ErrorHandler;
+  }
+
+  NeedReleaseRxTxBuffer = TRUE;
+
+  if ((Count != 1) || (Size < sizeof (EFI_FFA_PART_INFO_DESC))) {
+    Status = EFI_INVALID_PARAMETER;
+    DEBUG ((DEBUG_ERROR, "Invalid partition Info(%g). Count: %d, Size: %d\n", &gTpm2ServiceFfaGuid, Count, Size));
+    goto ErrorHandler;
+  }
+
+  TpmPartInfo = (EFI_FFA_PART_INFO_DESC *)RxBuffer;
+
+  ArmFfaLibRxRelease (PartId);
+  NeedReleaseRxTxBuffer = FALSE;
+
+  Status = Tpm2FfaCheckInterfaceVersion (TpmPartInfo->PartitionId);
+  if (EFI_ERROR (Status)) {
+    goto ErrorHandler;
+  }
+
+  /*
+   * Set the PcdTpmInstanceGuid to TPM using CRB over FF-A
+   * So that Tpm2DeviceLibRouterPei/Dxe can communicate with
+   * TPM secure partition.
+   */
+  Size   = sizeof (GUID);
+  Status = PcdSetPtrS (
+             PcdTpmInstanceGuid,
+             &Size,
+             &gTpm2ServiceFfaGuid
+             );
+  ASSERT_EFI_ERROR (Status);
+
+  Status = PeiServicesInstallPpi (&mTpmSelectedPpi);
+  ASSERT_EFI_ERROR (Status);
+
+  return EFI_SUCCESS;
+
+ErrorHandler:

```

Comment:
Can we just rename this to "Cleanup" or something generic?

Reply:
And in this symbol, just set the pcd to the outcome of this function, and install the ppi list of the outcome of this function.